### DISCUSSION TITLE
Struggling with installation on CentOS 7 (body of constexpr function ‘constexpr double opentime::v1_0::fabs(double)’ not a return-statement)

### DISCUSSION THREAD

**Discussion by OlivierArgentieri** [AcademySoftwareFoundation/OpenTimelineIO#1676] [Q&A]
## Environment

- CentOS 7
- GCC (tested with 6.5.0 and 12.3.0)
- CMake 3.28.20231017
- Python 3.10

## Context

I'm trying to install OpenTimelineIO from source on CentOS 7, like this:

(According to the documentation)
```sh
git clone https://github.com/AcademySoftwareFoundation/OpenTimelineIO.git
cd ./OpenTimelineIO
git checkout v0.15

mkdir build
cd build
cmake .. -DCMAKE_INSTAL_PREFIX=<PATH_TO_OPENTIMELINEIO>/install 
cmake --build . --target install --config Release
```

## Logs

```sh
[ 10%] Building CXX object src/opentime/CMakeFiles/opentime.dir/rationalTime.o
In file included from <OPENTIMELINEIO_SOURCE_DIR>/src/opentime/rationalTime.cpp:4:0:
<OPENTIMELINEIO_SOURCE_DIR>/src/opentime/rationalTime.h: In function ‘constexpr double opentime::v1_0::fabs(double)’:
<OPENTIMELINEIO_SOURCE_DIR>/src/opentime/rationalTime.h:32:1: error: body of constexpr function ‘constexpr double opentime::v1_0::fabs(double)’ not a return-statement
 }
 ^
gmake[2]: *** [src/opentime/CMakeFiles/opentime.dir/rationalTime.o] Error 1
gmake[1]: *** [src/opentime/CMakeFiles/opentime.dir/all] Error 2
gmake: *** [all] Error 2
```

## Things tried

```sh
git clone https://github.com/AcademySoftwareFoundation/OpenTimelineIO.git
cd ./OpenTimelineIO
git checkout v0.15 # i have also tried all above steps on main branch
```

then tried with c++11
```sh
cmake -S . -B ./build -DCMAKE_INSTALL_PREFIX=<OPENTIMELINEIO_SOURCE_DIR>/install \
 -DCMAKE_BUILD_TYPE=Release \
 -DCMAKE_CXX_STANDARD=11

cmake --build ./build --config Release --target install
```

Also in c++14 (just in case)
```sh
cmake -S . -B ./build -DCMAKE_INSTALL_PREFIX=<OPENTIMELINEIO_SOURCE_DIR>/install \
 -DCMAKE_BUILD_TYPE=Release \
 -DCMAKE_CXX_STANDARD=14

cmake --build ./build --config Release --target install
```

Also with ` -DOTIO_SHARED_LIBS=OFF -DOTIO_CXX_COVERAGE=ON` (according to the Ubuntu github action build command)

```sh
cmake -S . -B ./build -DCMAKE_INSTALL_PREFIX=<OPENTIMELINEIO_SOURCE_DIR>/install \
 -DCMAKE_BUILD_TYPE=Release \
 -DCMAKE_CXX_STANDARD=11
 -DOTIO_SHARED_LIBS=OFF -DOTIO_CXX_COVERAGE=ON

cmake --build ./build --config Release --target install
```

I also tried the same script as the GitHub action one 

```sh
cmake -E make_directory <OPENTIMELINEIO_SOURCE_DIR>/build
cd ./build
cmake <OPENTIMELINEIO_SOURCE_DIR> \
 -DCMAKE_INSTALL_PREFIX=<OPENTIMELINEIO_SOURCE_DIR>/install \
 -DOTIO_SHARED_LIBS=OFF -DOTIO_CXX_COVERAGE=ON

cmake --build . --config Release
```

But i always got this same error, do you have an idea of what i'm doing wrong ?
Maybe it's an environment issue ?

**Comment by darbyjohnston** [comment]
Hi, it looks like your commands are OK, but the error message is from an older compiler or C++ version. Is it possible that CMake is picking up an older compiler on your system? I don't have a CentOS 7 system but I am currently compiling v0.15 with g++ 11 on Ubuntu.

Also the OpenTimelineIO CMake file will set the C++ version to 14 by default.

**Reply by OlivierArgentieri** [reply]
Thank you for your message i will check this way !

**Comment by OlivierArgentieri** [comment]
Ok so i use [rez](https://github.com/AcademySoftwareFoundation/rez) to create a new environment with:
- gcc 12.3.0
- cmake 3.28
- mpfr 4.2.1
- mpc 1.3.1
- gmp 6.3.0

And i re-got the same error but thanks to @darbyjohnston for the idea because,
after adding `-DCMAKE_C_COMPILER=<GCC_SOURCE_DIR>/bin/gcc` to cmake option i can  now successfully compile without any errors OpenTimelineIO.

I know it's dirty, but that confirm the hypothesis about the wrong gcc version used in my environment.

Anyway, i will close this discussion, thanks for the help!
